### ISSUE TITLE
Add non-blocking option to cursors for changefeeds

### ISSUE THREAD

**Issue by larkost** [rethinkdb/rethinkdb#3529]
It would be a lot easier to work with `changefeeds` if they had some method of doing a (more or less) non-blocking read on them. For example, the following blocks unless something else adds a change:

```python
feed = r.table('smith').changes().run(conn)
next(feed) # blocks forever here without outside intervention
```

To avoid this we could mimic Python's `Queue` object and add a `.get([block, [timeout]])` method to the `Cursor` object that would either return an already fetched item, or do a short-timeout request for the next object(s) and if that got nothing raise an exception based on `Queue.Empty`. There is a similar model to follow for the Ruby driver in the `thread` module also named `Queue` that could be used the same way, and this is not an issue in JavaScript.

```python
feed = r.table('smith').changes().run(conn)
while True:
    try:
        item = feed.get(False)
        # work on item
    except Queue.Empty:
        # do something else
```

For completeness we could also add `get_nowait()` which is equivalent to `get(block=False, timeout=0)`.

This is slightly orthogonal but related to an number of other issues:
* Reading from multiple changefeeds in Python #3298
* Asynchronous Driver For Python <Tornado and Tulip> #2622
* Execute queries in parallel on a single connection #3296
* Consider basing our client protocol on an existing underlying stream protocol #3460

**Comment by gchpaco** [comment]
Ruby equivalent is #3678.

**Comment by danielmewes** [comment]
Moving to 2.0.

@larkost , @coffeemug , @timmaxw, @mlucy and me concluded in an offline discussion that we should have an optarg `cursor.next(wait=False)` to make the element retrieval non-blocking.
This would complement the native async support for reading from cursors.

If there's no element, `cursor.next(wait=False)` will throw an exception.

An open question is if you should be able to pass a timeout instead of `False`. I don't think it's necessary, but also don't oppose it if it's easy to do.

**Comment by coffeemug** [comment]
> An open question is if you should be able to pass a timeout instead of `False`.

I don't think this would be particularly useful. I'd keep that out and add it people demand it.

**Comment by mlucy** [comment]
I'd add it, personally: I've wanted it before and it isn't that hard.

**Comment by larkost** [comment]
Having a numeric value would be nice, and simplify some parts of testing (otherwise I will just wrap it like I do now).

**Comment by danielmewes** [comment]
Let's do it with a timeout then.

**Comment by danielmewes** [comment]
We should add this in Python and Ruby.

Assigning @Tryneus.

**Comment by Tryneus** [comment]
This has been added to the python driver in `next` in commit 3713807.  Review was 2621.  Will be in release 2.0.  Working on the ruby driver next.

**Comment by Tryneus** [comment]
This feature has been implemented for the Ruby driver and is up in review 2650.

**Comment by Tryneus** [comment]
The ruby driver implementation has been merged to `next` in commit 9ebc172.  Will be in release 2.0.
